File: src/main/kotlin/com/github/dungphan/unityindex/tools/unity/UnityGraphContextTool.kt
package com.github.dungphan.unityindex.tools.unity

import com.github.dungphan.unityindex.constants.ToolNames
import com.github.dungphan.unityindex.server.models.ContentBlock
import com.github.dungphan.unityindex.server.models.ToolCallResult
import com.github.dungphan.unityindex.tools.AbstractMcpTool
import com.github.dungphan.unityindex.tools.intelligence.GetDiagnosticsTool
import com.github.dungphan.unityindex.tools.models.DiagnosticSummary
import com.github.dungphan.unityindex.tools.models.EdgeWithEndpoint
import com.github.dungphan.unityindex.tools.models.FileStructureResult
import com.github.dungphan.unityindex.tools.models.GraphContextRequest
import com.github.dungphan.unityindex.tools.models.GraphContextResponse
import com.github.dungphan.unityindex.tools.models.GraphSnapshotRequest
import com.github.dungphan.unityindex.tools.models.NodeKind
import com.github.dungphan.unityindex.tools.models.ProblemInfo
import com.github.dungphan.unityindex.tools.navigation.FileStructureTool
import com.github.dungphan.unityindex.tools.schema.SchemaBuilder
import com.github.dungphan.unityindex.util.GraphTraversal
import com.github.dungphan.unityindex.util.UnityAssetGraphBuilder
import com.intellij.openapi.project.Project
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Day-6 MCP surface — see `docs/graph-mcp-tools.md` §3.4.
 * Single node + 1-hop neighborhood, optimized for LLM prompts. Optionally
 * enriches script nodes with a code summary (FileStructureTool) and
 * diagnostics (GetDiagnosticsTool). No new analysis paths.
 */
class UnityGraphContextTool : AbstractMcpTool() {

    companion object {
        private const val DEFAULT_MAX_NEIGHBORS = 50
        private const val HARD_MAX_NEIGHBORS = 10_000
    }

    override val requiresPsiSync: Boolean = false

    override val name: String = ToolNames.UNITY_GRAPH_CONTEXT

    override val description: String = """
        Return a single graph node plus its 1-hop neighborhood, optimized for LLM prompt construction.
        Optionally enriches script nodes with a code-summary (ide_file_structure) and diagnostics (ide_diagnostics).
    """.trimIndent()

    override val inputSchema: JsonObject = SchemaBuilder.tool()
        .projectPath()
        .stringProperty("node_id", "Graph node ID to focus on.", required = true)
        .booleanProperty("include_code_summary", "Default true. For script nodes, attaches a markdown-ish code summary.")
        .booleanProperty("include_diagnostics", "Default false. Attach diagnostics for the node's file (if any).")
        .intProperty("max_neighbors", "Cap per direction. Default $DEFAULT_MAX_NEIGHBORS.")
        .stringProperty("request_id", "Optional; echoed back on the response for client correlation.")
        .build()

    override suspend fun doExecute(project: Project, arguments: JsonObject): ToolCallResult {
        val request = try {
            json.decodeFromJsonElement(GraphContextRequest.serializer(), arguments)
        } catch (e: Exception) {
            return createErrorResult("Invalid arguments: ${e.message}")
        }
        val nodeId = request.node_id?.takeIf { it.isNotBlank() }
            ?: return createStructuredErrorResult(buildJsonObject {
                put("error", buildJsonObject {
                    put("kind", JsonPrimitive("invalid_id"))
                    put("detail", JsonPrimitive("node_id is required"))
                })
            })
        val includeCodeSummary = request.include_code_summary ?: true
        val includeDiagnostics = request.include_diagnostics ?: false
        val maxNeighbors = (request.max_neighbors ?: DEFAULT_MAX_NEIGHBORS).coerceIn(1, HARD_MAX_NEIGHBORS)

        return try {
            val snapshotReq = GraphSnapshotRequest(project_path = request.project_path)
            val fullResponse = withContext(Dispatchers.IO) {
                UnityAssetGraphBuilder.build(project, snapshotReq)
            }
            val adj = GraphTraversal.buildAdjacency(fullResponse.snapshot)
            val result = GraphTraversal.context(
                adj,
                nodeId,
                GraphTraversal.ContextOptions(maxNeighbors = maxNeighbors),
            ) ?: return createStructuredErrorResult(buildJsonObject {
                put("error", buildJsonObject {
                    put("kind", JsonPrimitive("invalid_id"))
                    put("detail", JsonPrimitive("node_id '$nodeId' not present in the current snapshot"))
                })
            })

            val incoming = result.incoming.map { EdgeWithEndpoint(it.edge, it.other) }
            val outgoing = result.outgoing.map { EdgeWithEndpoint(it.edge, it.other) }

            val codeSummary: String? = if (includeCodeSummary && result.node.kind == NodeKind.SCRIPT && result.node.path != null) {
                tryCodeSummary(project, result.node.path)
            } else null
            val diagnostics: List<DiagnosticSummary>? = if (includeDiagnostics && result.node.path != null) {
                tryDiagnostics(project, result.node.path).takeIf { it.isNotEmpty() }
            } else null

            val response = GraphContextResponse(
                request_id = request.request_id,
                generated_at = java.time.Instant.now().toString(),
                warnings = null,
                node = result.node,
                incoming = incoming,
                outgoing = outgoing,
                code_summary = codeSummary,
                diagnostics = diagnostics,
                truncated = if (result.truncated) true else null,
            )
            createJsonResult(response)
        } catch (e: IllegalStateException) {
            createErrorResult(e.message ?: "Failed to compute context")
        } catch (e: Exception) {
            createErrorResult("Failed to compute context: ${e.message}")
        }
    }

    private suspend fun tryCodeSummary(project: Project, filePath: String): String? {
        return try {
            val tool = FileStructureTool()
            val res = tool.execute(
                project,
                buildJsonObject { put("file", JsonPrimitive(filePath)) },
                skipPsiSync = true,
            )
            if (res.isError) return null
            val text = (res.content.firstOrNull() as? ContentBlock.Text)?.text ?: return null
            val parsed = try {
                json.decodeFromString(FileStructureResult.serializer(), text)
            } catch (_: Exception) {
                // FileStructureTool may return a wrapped tree string. Return raw text
                // (still useful) if it isn't structured JSON.
                return text.takeIf { it.isNotBlank() }
            }
            parsed.structure.takeIf { it.isNotBlank() }
        } catch (_: Exception) {
            null
        }
    }

    private suspend fun tryDiagnostics(project: Project, filePath: String): List<DiagnosticSummary> {
        return try {
            val tool = GetDiagnosticsTool()
            val res = tool.execute(
                project,
                buildJsonObject { put("file", JsonPrimitive(filePath)) },
                skipPsiSync = true,
            )
            if (res.isError) return emptyList()
            val text = (res.content.firstOrNull() as? ContentBlock.Text)?.text ?: return emptyList()
            val problems = extractProblems(text)
            problems.map { p ->
                DiagnosticSummary(
                    severity = p.severity,
                    message = p.message,
                    line = p.line,
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun extractProblems(text: String): List<ProblemInfo> {
        // GetDiagnosticsTool returns a structured payload that may be the raw
        // JSON, a wrapped envelope, or a formatted string depending on the
        // user's response-format setting. Try the raw JSON path first.
        return try {
            val element = json.parseToJsonElement(text)
            val obj = element as? JsonObject ?: return emptyList()
            val problemsArr = obj["problems"] ?: return emptyList()
            json.decodeFromJsonElement(kotlinx.serialization.builtins.ListSerializer(ProblemInfo.serializer()), problemsArr)
        } catch (_: Exception) {
            emptyList()
        }
    }
}

File: vscode-extension/src/tools/unity/unityGraphContextTool.ts
import type {
  ContextRequest,
  ContextResponse,
  DiagnosticSummary,
  SnapshotRequest,
} from "@unity-index/graph-core";
import { buildAdjacency, context } from "@unity-index/graph-core";
import { AbstractMcpTool, ToolContext } from "../abstractTool";
import { TOOL_NAMES, PARAM_NAMES } from "../../constants";
import { ToolCallResult } from "../../models/jsonRpc";
import { ProjectContext } from "../../server/projectResolver";
import { Args } from "../../utils/args";
import { SchemaBuilder } from "../../utils/schema";
import { buildAssetGraph } from "../../utils/unityAssetGraphBuilder";
import { FileStructureTool } from "../navigation/fileStructureTool";
import { GetDiagnosticsTool } from "../intelligence/getDiagnosticsTool";
import type {
  DiagnosticsResult,
  FileStructureItem,
  FileStructureResult,
} from "../../models/toolModels";

const DEFAULT_MAX_NEIGHBORS = 50;

/**
 * Day-6 MCP surface — see graph-mcp-tools.md §3.4.
 * Single node + 1-hop neighborhood, flattened for LLM prompts. For `script`
 * nodes, optionally pulls a code summary (delegates to FileStructureTool)
 * and diagnostics (delegates to GetDiagnosticsTool).
 */
export class UnityGraphContextTool extends AbstractMcpTool {
  protected readonly requiresLsp = false;
  readonly isHeavyScan = true;

  readonly name = TOOL_NAMES.UNITY_GRAPH_CONTEXT;
  readonly description =
    "Return a single graph node plus its 1-hop neighborhood, optimized for LLM prompt construction. " +
    "Optionally enriches script nodes with a code-summary (ide_file_structure) and diagnostics (ide_diagnostics).";
  readonly inputSchema = SchemaBuilder.tool()
    .projectPath()
    .stringProperty("node_id", "Graph node ID to focus on.", true)
    .booleanProperty(
      "include_code_summary",
      "Default true. For script nodes, attaches a markdown-ish code summary.",
    )
    .booleanProperty(
      "include_diagnostics",
      "Default false. Attach diagnostics for the node's file (if any).",
    )
    .intProperty(
      "max_neighbors",
      `Cap per direction. Default ${DEFAULT_MAX_NEIGHBORS}.`,
    )
    .stringProperty(
      "request_id",
      "Optional; echoed back on the response for client correlation.",
    )
    .build();

  protected async doExecute(
    project: ProjectContext,
    args: Args,
    ctx: ToolContext,
  ): Promise<ToolCallResult> {
    const request = args as unknown as ContextRequest;
    const nodeId = request.node_id;
    if (typeof nodeId !== "string" || nodeId.length === 0) {
      return this.structuredError({
        error: { kind: "invalid_id", detail: "node_id is required" },
      });
    }
    const includeCodeSummary = request.include_code_summary ?? true;
    const includeDiagnostics = request.include_diagnostics ?? false;
    const maxNeighbors = clamp(
      request.max_neighbors ?? DEFAULT_MAX_NEIGHBORS,
      1,
      10_000,
    );

    try {
      const index = await ctx.assetIndex.get(project);
      const snapshotReq: SnapshotRequest = { project_path: project.rootPath };
      const fullResponse = await buildAssetGraph(
        project.rootPath,
        index,
        snapshotReq,
        ctx.signal,
      );
      const adj = buildAdjacency(fullResponse.snapshot);
      const result = context(adj, nodeId, { maxNeighbors });
      if (!result) {
        return this.structuredError({
          error: {
            kind: "invalid_id",
            detail: `node_id '${nodeId}' not present in the current snapshot`,
          },
        });
      }

      const response: ContextResponse = {
        generated_at: new Date().toISOString(),
        node: result.node,
        incoming: result.incoming,
        outgoing: result.outgoing,
      };
      if (result.truncated) response.truncated = true;
      if (request.request_id !== undefined) response.request_id = request.request_id;

      if (
        includeCodeSummary &&
        result.node.kind === "script" &&
        result.node.path
      ) {
        const summary = await tryCodeSummary(project, result.node.path, ctx);
        if (summary) response.code_summary = summary;
      }
      if (includeDiagnostics && result.node.path) {
        const diags = await tryDiagnostics(project, result.node.path, ctx);
        if (diags && diags.length > 0) response.diagnostics = diags;
      }

      return this.json(response);
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      return this.error(`Failed to compute context: ${message}`);
    }
  }
}

async function tryCodeSummary(
  project: ProjectContext,
  filePath: string,
  ctx: ToolContext,
): Promise<string | undefined> {
  try {
    const tool = new FileStructureTool();
    const res = await tool.execute(
      project,
      { [PARAM_NAMES.FILE]: filePath } as unknown as Args,
      ctx,
    );
    if (res.isError) return undefined;
    const text = res.content?.[0]?.text;
    if (typeof text !== "string") return undefined;
    const parsed = JSON.parse(text) as FileStructureResult;
    if (!parsed.items || parsed.items.length === 0) return undefined;
    return renderStructure(parsed.items, 0);
  } catch {
    return undefined;
  }
}

function renderStructure(items: FileStructureItem[], depth: number): string {
  const out: string[] = [];
  for (const item of items) {
    const indent = "  ".repeat(depth);
    const detail = item.detail ? ` ${item.detail}` : "";
    out.push(`${indent}- ${item.kind} \`${item.name}\`${detail} (L${item.line})`);
    if (item.children && item.children.length > 0) {
      out.push(renderStructure(item.children, depth + 1));
    }
  }
  return out.join("\n");
}

async function tryDiagnostics(
  project: ProjectContext,
  filePath: string,
  ctx: ToolContext,
): Promise<DiagnosticSummary[] | undefined> {
  try {
    const tool = new GetDiagnosticsTool();
    const res = await tool.execute(
      project,
      { [PARAM_NAMES.FILE]: filePath } as unknown as Args,
      ctx,
    );
    if (res.isError) return undefined;
    const text = res.content?.[0]?.text;
    if (typeof text !== "string") return undefined;
    const parsed = JSON.parse(text) as DiagnosticsResult;
    if (!parsed.problems) return undefined;
    return parsed.problems.map((p) => {
      const severity: DiagnosticSummary["severity"] =
        p.severity === "error"
          ? "error"
          : p.severity === "warning"
            ? "warning"
            : "info";
      const entry: DiagnosticSummary = {
        severity,
        message: p.message,
      };
      if (typeof p.line === "number") entry.line = p.line;
      return entry;
    });
  } catch {
    return undefined;
  }
}

function clamp(n: number, lo: number, hi: number): number {
  if (!Number.isFinite(n)) return lo;
  return Math.max(lo, Math.min(hi, Math.trunc(n)));
}
